### PR TITLE
New Resource: `azurerm_dev_center_environment_type`

### PR DESCRIPTION
**Pull request by neil-yechenwei**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description
This PR is to support new resource azurerm_dev_center_environment_type.

API Reference: [azure-rest-api-specs/specification/devcenter/resource-manager/Microsoft.DevCenter/stable/2023-04-01/devcenter.json at main · Azure/azure-rest-api-specs (github.com)](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/devcenter/resource-manager/Microsoft.DevCenter/stable/2023-04-01/devcenter.json)

Overview: [What is Microsoft Dev Box? - Microsoft Dev Box | Microsoft Learn](https://learn.microsoft.com/en-gb/azure/dev-box/overview-what-is-microsoft-dev-box)

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
--- PASS: TestAccDevCenterEnvironmentType_basic (978.42s)
--- PASS: TestAccDevCenterEnvironmentType_requiresImport (832.94s)
--- PASS: TestAccDevCenterEnvironmentType_complete (544.53s)
--- PASS: TestAccDevCenterEnvironmentType_update (1001.75s)
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* New Resource: `azurerm_dev_center_environment_type`

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes https://github.com/hashicorp/terraform-provider-azurerm/issues/23367

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/devcenter/dev_center_environment_type_resource_test.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+			Config: r.complete(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccDevCenterEnvironmentType_update(t *testing.T) {
```
Comment (ms-zhenhua) [inline internal/services/devcenter/dev_center_environment_type_resource_test.go:63]: please use the following the steps to also test add/remove tags:

basic -> complete -> update -> basic

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @neil-yechenwei ,

Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

File: internal/services/devcenter/dev_center_environment_type_resource.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+				if response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(*id)
+				}
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			state := DevCenterEnvironmentTypeResourceModel{}
+			if model := resp.Model; model != nil {
+				state.Name = id.EnvironmentTypeName
+				state.DevCenterId = environmenttypes.NewDevCenterID(id.SubscriptionId, id.ResourceGroupName, id.DevCenterName).ID()
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_environment_type_resource.go:126]: Updated

File: internal/services/devcenter/dev_center_environment_type_resource_test.go (hunk 1 above)
Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_environment_type_resource_test.go:63]: Updated

**Comment by neil-yechenwei** [comment]
@ms-zhenhua , thanks for the comments. I updated PR. Please take another look.

File: internal/services/devcenter/dev_center_environment_type_resource.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/devcenter/dev_center_environment_type_resource.go]: `Name` is not updated

[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_environment_type_resource.go]: Updated

**Review by ms-zhenhua — APPROVED** [review]
LGTM~

File: website/docs/r/dev_center_environment_type.html.markdown [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+* `delete` - (Defaults to 30 minutes) Used when deleting this Dev Center Environment Type.
+* `read` - (Defaults to 5 minutes) Used when retrieving this Dev Center Environment Type.
+* `update` - (Defaults to 30 minutes) Used when updating this Dev Center Environment Type.
+
+## Import
+
+An existing Dev Center Environment Type can be imported into Terraform using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_dev_center_environment_type.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.DevCenter/devcenters/dc1/environmentTypes/et1
```
Comment (stephybun) [inline website/docs/r/dev_center_environment_type.html.markdown:73]: Is the casing here correct, should this be
```suggestion
terraform import azurerm_dev_center_environment_type.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.DevCenter/devCenters/dc1/environmentTypes/et1
```

**Review by stephybun — CHANGES_REQUESTED** [review]

File: internal/services/devcenter/dev_center_environment_type_resource_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+func (r DevCenterEnvironmentTypeTestResource) Exists(ctx context.Context, clients *clients.Client, state *pluginsdk.InstanceState) (*bool, error) {
+	id, err := environmenttypes.ParseDevCenterEnvironmentTypeID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := clients.DevCenter.V20230401.EnvironmentTypes.EnvironmentTypesGet(ctx, *id)
+	if err != nil {
+		return nil, fmt.Errorf("reading %s: %+v", *id, err)
```
Comment (stephybun) [inline internal/services/devcenter/dev_center_environment_type_resource_test.go:107]:
```suggestion
		return nil, fmt.Errorf("retrieving %s: %+v", *id, err)
```

File: internal/services/devcenter/dev_center_environment_type_resource_test.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	return utils.Bool(resp.Model != nil), nil
```
Comment (stephybun) [inline internal/services/devcenter/dev_center_environment_type_resource_test.go:110]:
```suggestion
	return pointer.To(resp.Model != nil), nil
```

Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_environment_type_resource_test.go]: Updated

File: internal/services/devcenter/dev_center_environment_type_resource_test.go (hunk 4 above)
Comment (neil-yechenwei) [reply internal/services/devcenter/dev_center_environment_type_resource_test.go]: Updated

File: website/docs/r/dev_center_environment_type.html.markdown (hunk 3 above)
Comment (neil-yechenwei) [reply website/docs/r/dev_center_environment_type.html.markdown]: Updated

**Comment by neil-yechenwei** [comment]
@stephybun , thanks for the comments. I updated PR. Please take another look. Below is the test result I just now triggered.

image

**Review by stephybun — APPROVED** [review]
Thanks @neil-yechenwei LGTM 👍

**Pull request merged**